cpp/src/arrow/type.h

```diff
@@ -191,16 +192,33 @@ class ARROW_EXPORT Fingerprintable {
 
 }  // namespace detail
 
+/// EXPERIMENTAL: Layout specification for a data type
 struct ARROW_EXPORT DataTypeLayout {
```

Comment:
@fsaintjacques IIRC, you requested something like the following changes some time ago?

Reply:
I did want something like this, but it doesn't exactly solve the issue. What I wanted to get rid of are the multiple `buffers[0]` with `0`, `1`, `2` scattered around where you have to know the index. DataTypeLayout is a step in automating this process.